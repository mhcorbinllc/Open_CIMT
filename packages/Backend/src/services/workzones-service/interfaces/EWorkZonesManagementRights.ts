import {convertDictionaryToViewLabelArray} from "../../../utils/convertDictionaryToViewLabelArray";

export enum EWorkZonesManagementRights {
  WORKZONES_VIEW = "WZM_V",
  WORKZONES_EDIT = "WZM_E",
}

const labelsDic: Record<EWorkZonesManagementRights, string>  = {
  [EWorkZonesManagementRights.WORKZONES_VIEW]: 'CIMs / View',
  [EWorkZonesManagementRights.WORKZONES_EDIT]: 'CIMs / Edit',
};

export const EWorkZonesManagementRightsArray = convertDictionaryToViewLabelArray(labelsDic);
