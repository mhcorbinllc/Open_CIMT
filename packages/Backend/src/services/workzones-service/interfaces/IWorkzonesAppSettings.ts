import {IWorkzone} from "./IWorkzone";
import {getDefaultAppItisCodes} from "./getDefaultAppItisCodes";
import {closureTypePresetValues} from "./closureTypePresetValues";
// App settings per company (not per device or user).

export interface IWorkzonesAppSettings {
  // See the apiWorkzonesAppSettingPost.ts's validation for the valid options per property

  deviceId: string;           // PK, "*" is for all devices of the company

  serviceChannel: string;
  psid: string;               // Courier font type
  txmode: EWZTXMode;
  intervalInMs: number;
  priority: number;           // 0..63
  useSignature: boolean;
  useEncryption: boolean;
  closureSelectionMode: EWZClosureSelectionMode;

  itisCodes: IItisCode[];

  createdAt: number;
  updatedAt: number;
  deletedAt: number;
}

export enum EWZTXMode {
  CONT = "CONT",
  ALT = "ALT",
}

export enum EWZClosureSelectionMode {
  CUSTOM = "CUSTOM",
  SIMPLE = "SIMPLE",
}

export interface IItisCode {
  code: number;
  label: string;
}

export const defaultWorkzoneAppSettings: IWorkzonesAppSettings = {
  deviceId: '',
  serviceChannel: "172",
  psid: '8002',
  txmode: EWZTXMode.CONT,
  intervalInMs: 1000,
  priority: 4,
  useSignature: false,
  useEncryption: false,
  closureSelectionMode: EWZClosureSelectionMode.SIMPLE,
  itisCodes: getDefaultAppItisCodes(),
  createdAt: 0,
  updatedAt: 0,
  deletedAt: 0,
};

export interface IValidationErrorItisRemoveBlockedByWorkzone {
  removedItisCodes: number[];
  workzone: IWorkzone;
}

export interface IClosureTypeRemovedItisCodes {
  closureType: string;
  removedItisCodes: number[];
}

export const getRemovedSimpleClosureTypeItisCodes  = (
  itisCodes: IItisCode[],
): IClosureTypeRemovedItisCodes[] => {
  const closureTypeRemovedItisCodes: IClosureTypeRemovedItisCodes[] = [];
  const codesList = itisCodes.map(itisCode => (itisCode.code));
  Object.keys(closureTypePresetValues)
    .forEach((key) => {
      const newRecord: IClosureTypeRemovedItisCodes = {
        closureType: key,
        removedItisCodes: [],
      };
      closureTypePresetValues[key].itisCodes
        .forEach((cti: number) => {
          const exists = codesList.includes(cti);
          if (exists) return;
          newRecord.removedItisCodes.push(cti);
        });
      closureTypeRemovedItisCodes.push(newRecord);
    });

  return closureTypeRemovedItisCodes.filter(closureType => closureType.removedItisCodes.length > 0);
};
