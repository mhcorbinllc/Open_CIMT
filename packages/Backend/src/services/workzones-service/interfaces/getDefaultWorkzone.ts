import * as moment from "moment";

import {
  EWZBroadcastSelectionMode,
  EWZClosureType,
  EWZLane,
  EWZShoulder,
  IWorkzone,
} from "./IWorkzone";
import {
  EWZClosureSelectionMode,
} from "./IWorkzonesAppSettings";

export const getDefaultWorkzone = (): IWorkzone => {
  const now = moment().seconds(0);

  return {
    workzoneId: '',
    kapschId: -1,
    name: '',
    active: true,

    notes: '',
    start: now.toDate(),
    end: now.clone().add(3, 'days')
      .toDate(),
    forEver: false,

    closureSelectionMode: EWZClosureSelectionMode.SIMPLE,
    closureType: EWZClosureType.NONE,
    itisCodes: [],
    closedLane: EWZLane.NONE,
    closedShoulder: EWZShoulder.NONE,
    speedLimitInMiles: 0,

    closedLaneWidthInMeters: 0,
    workersPresent: false,

    point: {
      lat: 0,
      lng: 0,
      alt: 0,
    },

    path: {points: []},

    heading: [],

    broadcast: {
      selectionMode: EWZBroadcastSelectionMode.REGION,
      selectionByRegion: {region: {points: []}},
      selectionByRadius: {radiusInMeters: 0},
    },

    createdAt: 0,
    updatedAt: 0,
    deletedAt: 0,
  };
};

