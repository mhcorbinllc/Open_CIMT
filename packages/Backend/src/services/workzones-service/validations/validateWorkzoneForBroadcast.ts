import {
  IWorkzone,
  EWZBroadcastSelectionMode,
  IGeoPoint,
  IGeoLine,
  IGeoPolygon,
  EWZClosureType,
  EWZLane,
  EWZShoulder,
  EWZHeading,
  EWZClosureTypeDeprecated,
} from "../interfaces/IWorkzone";

import {EWZClosureSelectionMode} from "../interfaces/IWorkzonesAppSettings";

import {validateWorkzoneForUpdate} from "./validateWorkzoneForUpdate";

import {IValidateDataEngineRules} from "core-library/dist/commonJs/validation-engine";
import {validateWorkzoneElements} from "../interfaces/validateWorkzoneElements";
import {validateDataMethods} from "core-library/dist/commonJs/validation-engine";

// This is the complete validation of a CIM to be possible to broadcast it.
// Here we implement the validation that Kapsch needs for a CIM.
// Developer, remember to OVERLOAD (and not override) the validateWorkzoneForUpdate's validations.

// Dev notes: Type of closure data: https://docs.google.com/spreadsheets/d/1_y0YbEvzRtHp4KFFruyA3_xVY-RTKwM_5EfnrR0r1M0/edit?usp=sharing

export const validateWorkzoneForBroadcast: IValidateDataEngineRules<IWorkzone> = {
  ...validateWorkzoneForUpdate,

  active: (active: boolean, workzone) => {
    const parentValidation = validateWorkzoneForUpdate.active(active, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    if (!active) return 'Cannot broadcast inactive CIM!';
    return true;
  },

  end: (end_: Date, workzone) => {
    end_; // 4TS
    if (workzone.forEver) return true;

    // End validation
    const start = new Date(workzone.start);
    const end = new Date(workzone.end);
    if (end.valueOf() <= start.valueOf()) return 'The "End" date/time cannot be before or the same as the Start date';
    if (end.valueOf() <= Date.now()) return 'The "End" date/time is before the current time, nothing will be broadcast';

    // Minutes duration validation
    const startDate = new Date(workzone.start);
    const endDate = new Date(workzone.end);
    // Dev note: If the number of the days changed, consider updating also the point: #202108190832
    const maxDuration = 32000;
    const duration: number = Math.round((endDate.getTime() - startDate.getTime()) / 1000) / 60;
    if (duration > maxDuration) return 'Minutes duration from start to end date exceeds the maximum number of minutes allowed (22 days, 5 hours, and 20 minutes or 32000 minutes)';

    return true;
  },

  speedLimitInMiles: (speedLimitInMiles: number, workzone) => {
    const parentValidation = validateWorkzoneForUpdate.speedLimitInMiles(speedLimitInMiles, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    if (workzone.closureSelectionMode === EWZClosureSelectionMode.CUSTOM) return true;
    const hasSpeedLimit = speedLimitInMiles > 0;
    if (hasSpeedLimit && speedLimitInMiles < 15) return 'Speed limit must be 15mph or greater';
    if (speedLimitInMiles > 70) return 'Speed limit cannot be more than 70mph';
    if (hasSpeedLimit && speedLimitInMiles % 5 !== 0) return 'Speed limit must be in increments of 5mph.';
    return true;
  },

  closureType: (closureType: EWZClosureType, workzone) => {
    const parentValidation = validateWorkzoneForUpdate.closureType(closureType, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    const isSimple = workzone.closureSelectionMode === EWZClosureSelectionMode.SIMPLE;
    if (
      isSimple
      && closureType === EWZClosureType.NONE
    ) {
      return 'Closure type is not selected';
    }
    if (
      isSimple
      && validateDataMethods.isEnumValue(closureType, EWZClosureTypeDeprecated)
    ) {
      return 'Closure type is deprecated, please select a different closure type';
    }
    return true;
  },

  itisCodes: (itisCodes: number[], workzone) => {
    const parentValidation = validateWorkzoneForUpdate.itisCodes(itisCodes, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;

    if (
      workzone.closureSelectionMode === EWZClosureSelectionMode.SIMPLE
      && workzone.closureType === EWZClosureType.NONE
    ) return true; // The closureType validation is shown, don't complain about the ITIS codes.

    if (
      workzone.closureSelectionMode === EWZClosureSelectionMode.SIMPLE
      && !!workzone.closureType
      && workzone.closureType !== EWZClosureType.NONE
      && itisCodes.length === 0
    ) return 'The selected closure type has not returned any ITIS codes';

    if (itisCodes.length === 0) return 'No ITIS code is selected';

    return true;
  },

  closedLane: (closedLane: EWZLane, workzone) => {
    const parentValidation = validateWorkzoneForUpdate.closedLane(closedLane, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    if (workzone.closureSelectionMode === EWZClosureSelectionMode.CUSTOM) return true; // Controlled by simple closures only
    if (
      workzone.closureSelectionMode === EWZClosureSelectionMode.SIMPLE
      && workzone.closureType !== EWZClosureType.NONE
      && [
        EWZClosureType.LEFT_LANE_CLOSED_MERGE_RIGHT,
        EWZClosureType.RIGHT_LANE_CLOSED_MERGE_LEFT,
      ].includes(workzone.closureType as EWZClosureType)
      && closedLane === EWZLane.NONE
    ) {
      return 'Closed lane is not selected';
    }
    return true;
  },

  closedLaneWidthInMeters: (value: number, workzone) => {
    const parentValidation = validateWorkzoneForUpdate.closedLaneWidthInMeters(value, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    if (workzone.closureSelectionMode === EWZClosureSelectionMode.CUSTOM) return true;
    if (value < 0.3048) return "Should be bigger than 1 feet";
    if (value > 49.9872) return "Should be less than 164 feet";
    return true;
  },

  closedShoulder: (closedShoulder: EWZShoulder, workzone) => {
    const parentValidation = validateWorkzoneForUpdate.closedShoulder(closedShoulder, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    if (workzone.closureSelectionMode === EWZClosureSelectionMode.CUSTOM) return true;// Only simple type sets the closedShoulder
    if (
      workzone.closureSelectionMode === EWZClosureSelectionMode.SIMPLE
      && workzone.closureType !== EWZClosureType.NONE
      && [
        EWZClosureType.LEFT_SHOULDER_CLOSED,
        EWZClosureType.RIGHT_SHOULDER_CLOSED,
      ].includes(workzone.closureType as EWZClosureType)
      && closedShoulder === EWZShoulder.NONE
    ) return 'Closed shoulder is not selected';
    return true;
  },

  point: (point: IGeoPoint, workzone) => {
    const parentValidation = validateWorkzoneForUpdate.point(point, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    if (point.lat === 0 || point.lng === 0) return 'Reference point is not defined';
    return '';
  },

  path: (path: IGeoLine, workzone) => {
    const parentValidation = validateWorkzoneForUpdate.path(path, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    if (path.points.length <= 1) return 'Path is not defined';
    if (path.points.length > 63) return 'The maximum number of points allowed for the path is 63';
    return '';
  },

  heading: (headings: EWZHeading[], workzone) => {
    const parentValidation = validateWorkzoneForUpdate.heading(headings, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    if (headings.length > 16) return 'Too many headings provided';
    const duplicates = validateDataMethods.hasUniqueValues(headings);
    if (duplicates !== true) return 'There are duplicate headings provided';
    return '';
  },

  "broadcast.selectionByRegion.region": (region: IGeoPolygon, workzone) => {
    const parentValidation = validateWorkzoneForUpdate["broadcast.selectionByRegion.region"](region, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    if (workzone.broadcast.selectionMode === EWZBroadcastSelectionMode.RADIUS) return true;
    return validateWorkzoneElements.isValidGeoAreaPolygon(region);
  },

  "broadcast.selectionByRadius.radiusInMeters": (radius: number, workzone) => {
    const parentValidation = validateWorkzoneForUpdate["broadcast.selectionByRadius.radiusInMeters"](radius, workzone);
    if (!validateDataMethods.isValid(parentValidation)) return parentValidation;
    if (workzone.broadcast.selectionMode === EWZBroadcastSelectionMode.REGION) return true;
    if (radius === 0) return 'Radius broadcast area cannot be zero';
    return true;
  },
};
