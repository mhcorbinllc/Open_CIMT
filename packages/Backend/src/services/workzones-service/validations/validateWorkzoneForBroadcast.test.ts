import * as moment from "moment";

import {
  IWorkzone,
  EWZBroadcastSelectionMode,
  EWZLane,
  EWZShoulder,
  EWZHeading,
  EWZClosureType,
  EWZClosureTypeDeprecated,
} from "../interfaces/IWorkzone";

import {getDefaultWorkzone} from "../interfaces/getDefaultWorkzone";
import {validationEngine} from "core-library/dist/commonJs/validation-engine";

import {validateWorkzoneForBroadcast} from "./validateWorkzoneForBroadcast";
import {EWZClosureSelectionMode} from "../interfaces/IWorkzonesAppSettings";

const now = Date.now();

const validBasicWorkzone: IWorkzone = {
  ...getDefaultWorkzone(),
  active: true,
  start: new Date(now + 200),
  end: new Date(now + 400),
  name: 'Demo CIM',
  workzoneId: 'ed-98467204',
  speedLimitInMiles: 0,
  itisCodes: [200, 300],
  closureType: EWZClosureType.LEFT_LANE_CLOSED_MERGE_RIGHT,
  closedLane: EWZLane.CENTER,
  closedShoulder: EWZShoulder.LEFT,
  closedLaneWidthInMeters: 12,
  point: {
    lng: 0.224423,
    lat: 0.134211,
  },
  heading: [EWZHeading.NORTH],
  path: {
    points: [
      {
        lng: 0.224423,
        lat: 0.134211,
      },
      {
        lng: 0.224423,
        lat: 0.134212,
      },
    ],
  },
  broadcast: {
    selectionMode: EWZBroadcastSelectionMode.RADIUS,
    selectionByRadius: {radiusInMeters: 12},
    selectionByRegion: {region: {points: []}},
  },
};

describe('validateWorkzoneForBroadcast', () => {
  test('Correct CIM', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('Empty CIM', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(getDefaultWorkzone()));

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(Object.keys(validation).length).not.toBe(0);
    expect(validation).toMatchSnapshot();
  });

  test('CIM with unselected closure type', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify({
      ...validBasicWorkzone,
      closureSelectionMode: EWZClosureSelectionMode.SIMPLE,
      closureType: EWZClosureType.NONE,
    } as IWorkzone));

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with selected closure type but no ITIS codes', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify({
      ...validBasicWorkzone,
      closureSelectionMode: EWZClosureSelectionMode.SIMPLE,
      closureType: EWZClosureType.LEFT_LANE_CLOSED_MERGE_RIGHT,
      itisCodes: [],
    } as IWorkzone));

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with custom closure selection mode and no ITIS codes', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify({
      ...validBasicWorkzone,
      closureSelectionMode: EWZClosureSelectionMode.CUSTOM,
      closureType: EWZClosureType.LEFT_LANE_CLOSED_MERGE_RIGHT,
      itisCodes: [],
    } as IWorkzone));

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with custom closure selection mode, no closureType and no ITIS codes', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify({
      ...validBasicWorkzone,
      closureSelectionMode: EWZClosureSelectionMode.CUSTOM,
      closureType: EWZClosureType.NONE,
      itisCodes: [],
    } as IWorkzone));

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with radius broadcast selection, without radius', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.broadcast.selectionMode = EWZBroadcastSelectionMode.RADIUS;
    workzone.broadcast.selectionByRadius.radiusInMeters = 0;

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with region broadcast selection, without points', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.broadcast.selectionMode = EWZBroadcastSelectionMode.REGION;
    workzone.broadcast.selectionByRegion.region.points = [];

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with region broadcast selection, with 1 point', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.broadcast.selectionMode = EWZBroadcastSelectionMode.REGION;
    workzone.broadcast.selectionByRegion.region.points = [
      {
        lat: 0.23,
        lng: 0.45,
      },
    ];

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with region broadcast selection, with 2 points', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.broadcast.selectionMode = EWZBroadcastSelectionMode.REGION;
    workzone.broadcast.selectionByRegion.region.points = [
      {
        lat: 0.23,
        lng: 0.45,
      },
      {
        lat: 0.23,
        lng: 0.46,
      },
    ];

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with region broadcast selection, with 3 points (does not close)', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.broadcast.selectionMode = EWZBroadcastSelectionMode.REGION;
    workzone.broadcast.selectionByRegion.region.points = [
      {
        lat: 0.23,
        lng: 0.45,
      },
      {
        lat: 0.23,
        lng: 0.46,
      },
      {
        lat: 0.24,
        lng: 0.46,
      },
    ];

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with region broadcast selection, with 3 points (close)', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.broadcast.selectionMode = EWZBroadcastSelectionMode.REGION;
    workzone.broadcast.selectionByRegion.region.points = [
      {
        lat: 0.23,
        lng: 0.45,
      },
      {
        lat: 0.23,
        lng: 0.46,
      },
      {
        lat: 0.23,
        lng: 0.45,
      },
    ];

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with region broadcast selection, with 4 points (does not close)', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.broadcast.selectionMode = EWZBroadcastSelectionMode.REGION;
    workzone.broadcast.selectionByRegion.region.points = [
      {
        lat: 0.23,
        lng: 0.45,
      },
      {
        lat: 0.23,
        lng: 0.46,
      },
      {
        lat: 0.24,
        lng: 0.46,
      },
      {
        lat: 0.23,
        lng: 0.47,
      },
    ];

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with region broadcast selection, with 4 points (close)', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.broadcast.selectionMode = EWZBroadcastSelectionMode.REGION;
    workzone.broadcast.selectionByRegion.region.points = [
      {
        lat: 0.23,
        lng: 0.45,
      },
      {
        lat: 0.23,
        lng: 0.46,
      },
      {
        lat: 0.24,
        lng: 0.46,
      },
      {
        lat: 0.23,
        lng: 0.45,
      },
    ];

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM duration in minutes cannot be greater than 32000', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.start = new Date();
    workzone.end = moment(workzone.start).add(32001, 'minutes')
      .toDate();

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with duplicate headings', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.heading = [EWZHeading.NORTH, EWZHeading.NORTH];

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM speedLimitInMiles not divisible by 5mph', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.speedLimitInMiles = 19;

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM forever with wrong end date, should not complain', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.end = new Date(workzone.start.valueOf() - 100);
    workzone.forEver = true;

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with deprecated simple closure type fails validation', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.closureSelectionMode = EWZClosureSelectionMode.SIMPLE;
    workzone.closureType = EWZClosureTypeDeprecated.LEFT_2_LANES_CLOSED_AHEAD;

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM with deprecated simple closure type does not fail when custom mode is selected', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.closureSelectionMode = EWZClosureSelectionMode.CUSTOM;
    workzone.closureType = EWZClosureTypeDeprecated.LEFT_2_LANES_CLOSED_AHEAD;

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM closedLane and closedShoudler validations do not fail on custom selection type', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.closureSelectionMode = EWZClosureSelectionMode.CUSTOM;
    workzone.closedShoulder = EWZShoulder.NONE;
    workzone.closedLane = EWZLane.NONE;

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });

  test('CIM closedLane and closedShoudler validations do not fail on custom selection type with simple closure set to a reduced speed closure', () => {
    const workzone: IWorkzone = JSON.parse(JSON.stringify(validBasicWorkzone));
    workzone.closureSelectionMode = EWZClosureSelectionMode.CUSTOM;
    workzone.closureType = EWZClosureType.REDUCED_SPEED_20_MPH;
    workzone.closedShoulder = EWZShoulder.NONE;
    workzone.closedLane = EWZLane.NONE;

    const validation = validationEngine(workzone, validateWorkzoneForBroadcast);

    expect(validation).toMatchSnapshot();
  });
});
