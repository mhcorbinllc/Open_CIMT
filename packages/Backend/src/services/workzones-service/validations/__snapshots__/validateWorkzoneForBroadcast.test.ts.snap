// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateWorkzoneForBroadcast CIM closedLane and closedShoudler validations do not fail on custom selection type 1`] = `
Object {
  "dataValidation": Object {},
  "isValid": true,
  "messages": Array [],
}
`;

exports[`validateWorkzoneForBroadcast CIM closedLane and closedShoudler validations do not fail on custom selection type with simple closure set to a reduced speed closure 1`] = `
Object {
  "dataValidation": Object {},
  "isValid": true,
  "messages": Array [],
}
`;

exports[`validateWorkzoneForBroadcast CIM duration in minutes cannot be greater than 32000 1`] = `
Object {
  "dataValidation": Object {
    "end": "Minutes duration from start to end date exceeds the maximum number of minutes allowed (22 days, 5 hours, and 20 minutes or 32000 minutes)",
  },
  "isValid": false,
  "messages": Array [
    "End: Minutes duration from start to end date exceeds the maximum number of minutes allowed (22 days, 5 hours, and 20 minutes or 32000 minutes)",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM forever with wrong end date, should not complain 1`] = `
Object {
  "dataValidation": Object {},
  "isValid": true,
  "messages": Array [],
}
`;

exports[`validateWorkzoneForBroadcast CIM speedLimitInMiles not divisible by 5mph 1`] = `
Object {
  "dataValidation": Object {
    "speedLimitInMiles": "Speed limit must be in increments of 5mph.",
  },
  "isValid": false,
  "messages": Array [
    "Speed limit in miles: Speed limit must be in increments of 5mph.",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with custom closure selection mode and no ITIS codes 1`] = `
Object {
  "dataValidation": Object {
    "itisCodes": "No ITIS code is selected",
  },
  "isValid": false,
  "messages": Array [
    "Itis codes: No ITIS code is selected",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with custom closure selection mode, no closureType and no ITIS codes 1`] = `
Object {
  "dataValidation": Object {
    "itisCodes": "No ITIS code is selected",
  },
  "isValid": false,
  "messages": Array [
    "Itis codes: No ITIS code is selected",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with deprecated simple closure type does not fail when custom mode is selected 1`] = `
Object {
  "dataValidation": Object {},
  "isValid": true,
  "messages": Array [],
}
`;

exports[`validateWorkzoneForBroadcast CIM with deprecated simple closure type fails validation 1`] = `
Object {
  "dataValidation": Object {
    "closureType": "Closure type is deprecated, please select a different closure type",
  },
  "isValid": false,
  "messages": Array [
    "Closure type: Closure type is deprecated, please select a different closure type",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with duplicate headings 1`] = `
Object {
  "dataValidation": Object {
    "heading": "There are duplicate headings provided",
  },
  "isValid": false,
  "messages": Array [
    "Heading: There are duplicate headings provided",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with radius broadcast selection, without radius 1`] = `
Object {
  "dataValidation": Object {
    "broadcast.selectionByRadius.radiusInMeters": "Radius broadcast area cannot be zero",
  },
  "isValid": false,
  "messages": Array [
    "Broadcast/selection by radius/radius in meters: Radius broadcast area cannot be zero",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with region broadcast selection, with 1 point 1`] = `
Object {
  "dataValidation": Object {
    "broadcast.selectionByRegion.region": "Not enough points for a polygon",
  },
  "isValid": false,
  "messages": Array [
    "Broadcast/selection by region/region: Not enough points for a polygon",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with region broadcast selection, with 2 points 1`] = `
Object {
  "dataValidation": Object {
    "broadcast.selectionByRegion.region": "Not enough points for a polygon",
  },
  "isValid": false,
  "messages": Array [
    "Broadcast/selection by region/region: Not enough points for a polygon",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with region broadcast selection, with 3 points (close) 1`] = `
Object {
  "dataValidation": Object {
    "broadcast.selectionByRegion.region": "Not enough points for a polygon",
  },
  "isValid": false,
  "messages": Array [
    "Broadcast/selection by region/region: Not enough points for a polygon",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with region broadcast selection, with 3 points (does not close) 1`] = `
Object {
  "dataValidation": Object {
    "broadcast.selectionByRegion.region": "Not enough points for a polygon",
  },
  "isValid": false,
  "messages": Array [
    "Broadcast/selection by region/region: Not enough points for a polygon",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with region broadcast selection, with 4 points (close) 1`] = `
Object {
  "dataValidation": Object {},
  "isValid": true,
  "messages": Array [],
}
`;

exports[`validateWorkzoneForBroadcast CIM with region broadcast selection, with 4 points (does not close) 1`] = `
Object {
  "dataValidation": Object {
    "broadcast.selectionByRegion.region": "The last point should match the start point to close the polygon",
  },
  "isValid": false,
  "messages": Array [
    "Broadcast/selection by region/region: The last point should match the start point to close the polygon",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with region broadcast selection, without points 1`] = `
Object {
  "dataValidation": Object {
    "broadcast.selectionByRegion.region": "Not enough points for a polygon",
  },
  "isValid": false,
  "messages": Array [
    "Broadcast/selection by region/region: Not enough points for a polygon",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with selected closure type but no ITIS codes 1`] = `
Object {
  "dataValidation": Object {
    "itisCodes": "The selected closure type has not returned any ITIS codes",
  },
  "isValid": false,
  "messages": Array [
    "Itis codes: The selected closure type has not returned any ITIS codes",
  ],
}
`;

exports[`validateWorkzoneForBroadcast CIM with unselected closure type 1`] = `
Object {
  "dataValidation": Object {
    "closureType": "Closure type is not selected",
  },
  "isValid": false,
  "messages": Array [
    "Closure type: Closure type is not selected",
  ],
}
`;

exports[`validateWorkzoneForBroadcast Correct CIM 1`] = `
Object {
  "dataValidation": Object {},
  "isValid": true,
  "messages": Array [],
}
`;

exports[`validateWorkzoneForBroadcast Empty CIM 1`] = `
Object {
  "dataValidation": Object {
    "broadcast.selectionByRegion.region": "Not enough points for a polygon",
    "closedLaneWidthInMeters": "Should be bigger than 1 feet",
    "closureType": "Closure type is not selected",
    "name": "Length must be between 2 and 300 characters",
    "path": "Path is not defined",
    "point": "Reference point is not defined",
    "workzoneId": "Required",
  },
  "isValid": false,
  "messages": Array [
    "Workzone id: Required",
    "Name: Length must be between 2 and 300 characters",
    "Closure type: Closure type is not selected",
    "Closed lane width in meters: Should be bigger than 1 feet",
    "Point: Reference point is not defined",
    "Path: Path is not defined",
    "Broadcast/selection by region/region: Not enough points for a polygon",
  ],
}
`;
